From c72b79a6b9818d65546cce5cc2918912412de577 Mon Sep 17 00:00:00 2001
From: Marko Lindqvist <cazfi74@gmail.com>
Date: Mon, 1 Aug 2022 06:35:13 +0300
Subject: [PATCH 27/49] do_attack(): Shrink city only after complete removal of
 the defender

Destruction of the city could wipe the defender too early, if it
was supported by that city.

See osdn #45240

Signed-off-by: Marko Lindqvist <cazfi74@gmail.com>
---
 server/unithand.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/server/unithand.c b/server/unithand.c
index 359671d67d..981fffb773 100644
--- a/server/unithand.c
+++ b/server/unithand.c
@@ -4843,7 +4843,6 @@ static bool do_attack(struct unit *punit, struct tile *def_tile,
   char attacker_fp[MAX_LEN_LINK], defender_fp[MAX_LEN_LINK];
   char attacker_tired[MAX_LEN_LINK];
   struct unit *ploser, *pwinner;
-  struct city *pcity;
   int moves_used, def_moves_used; 
   int old_unit_vet, old_defender_vet, vet;
   int winner_id;
@@ -4953,10 +4952,6 @@ static bool do_attack(struct unit *punit, struct tile *def_tile,
                              unit_owner(pdefender),
                              def_tile, unit_link(pdefender));
 
-  if (pdefender->hp <= 0
-      && (pcity = tile_city(def_tile))) {
-    unit_attack_civilian_casualties(punit, pcity, paction, "attack");
-  }
   if (punit->hp > 0 && pdefender->hp > 0) {
     /* Neither died */
     send_combat(punit, pdefender, punit->veteran - old_unit_vet,
@@ -5040,6 +5035,7 @@ static bool do_attack(struct unit *punit, struct tile *def_tile,
     wipe_unit(ploser, ULR_KILLED, unit_owner(pwinner));
   } else {
     /* The defender lost, the attacker punit lives! */
+    struct city *pcity;
 
     log_debug("Defender lost: %s %s against %s %s.",
               nation_rule_name(nation_of_player(pplayer)),
@@ -5097,6 +5093,14 @@ static bool do_attack(struct unit *punit, struct tile *def_tile,
     punit->moved = TRUE;	/* We moved */
     kill_unit(pwinner, ploser,
               vet && !utype_is_consumed_by_action(paction, punit->utype));
+
+    /* Now that dead defender is certainly no longer listed as unit
+     * supported by the city, we may even remove the city
+     * (if it shrinks from size 1) */
+    pcity = tile_city(def_tile);
+    if (pcity != NULL) {
+      unit_attack_civilian_casualties(punit, pcity, paction, "attack");
+    }
     if (unit_is_alive(winner_id)) {
       if (utype_is_consumed_by_action(paction, pwinner->utype)) {
         return TRUE;
-- 
2.35.1

